### ISSUE TITLE
Unable to add keybinds to nvim-tree with user/settings.lua

### ISSUE THREAD

**Issue by throwaway69420-69420** [kabinspace/AstroVim#79]
I have added this to the user/settings.lua
`  
overrides = {

    nvim_tree = {
      hijack_cursor = true,
      hijack_directories = {
        enable = true,
        auto_open = true,
      },
      view = {
        mappings = {
          custom_only = false,
          list = {
            { key = { "l", "<CR>", "o" }, cb = tree_cb "edit" },
            { key = "h", cb = tree_cb "close_node" },
            { key = "v", cb = tree_cb "vsplit" },
          },
        },
      },
    },
},
`

and this added at the top
`local tree_cb = require'nvim-tree.config'.nvim_tree_callback
`

the keybinds don't get used though, even after a :PackerCompile 
it gets used if I do the same in lua/configs/nvim-tree.lua

**Comment by osted** [comment]
In my installation user/settings.lua is not executed at all until I do it manually (like :source ~/.config/nvim/lua/user/settings.lua ). Only then my modifications are applied. 

I might remember wrong, but there was an earlier version of astrovim which explicitly invoked settings.lua from init.lua and this changed couple of days ago?

**Comment by osted** [comment]
Unfortunately that does not work for me. In user/settings.lua I have this line:

`set.whichwrap = '<>[]'`

When I load nvim my cursor does not wrap over beginning / end of line, then I type: 

```
:PackerSync
:PackerCompile
```
And it still does not wrap.

Then I type `:source ~/.config/nvim/lua/user/settings.lua` and immediately my cursor behaves as expected.

**Comment by kabinspace** [comment]
@throwaway69420-69420 Can you try adding those keybindings over [here]( https://github.com/kabinspace/AstroVim/blob/1ea38e912ac50f1b8640ea172dcc5a8bf67bde88/lua/user/settings.lua#L55) instead?

**Comment by kabinspace** [comment]
> Unfortunately that does not work for me. In user/settings.lua I have this line:
> 
> `set.whichwrap = '<>[]'`
> 
> When I load nvim my cursor does not wrap over beginning / end of line, then I type:
> 
> ```
> :PackerSync
> :PackerCompile
> ```
> 
> And it still does not wrap.
> 
> Then I type `:source ~/.config/nvim/lua/user/settings.lua` and immediately my cursor behaves as expected.

That's quite strange, are you sure you are on the latest version of astrovim? Also can you tell me which OS are you using?

**Comment by osted** [comment]
I am happy to report that after running `:AstroUpdate` settings.lua now execute without the need to manually source it.

I thought I was on the latest version, but must've missed something. Just FYI - I am running this in WSL2 / Ubuntu 21.10.

**Comment by throwaway69420-69420** [comment]
@kabinspace I am not sure how to add nvim-tree keybinds at [here](https://github.com/kabinspace/AstroVim/blob/1ea38e912ac50f1b8640ea172dcc5a8bf67bde88/lua/user/settings.lua#L55) 

`:h nvim-tree-mappings` doesn't show me anything useful for this case

**Comment by kabinspace** [comment]
@throwaway69420-69420 What are the commands that you want to map?

**Comment by throwaway69420-69420** [comment]
@kabinspace they are in my original post inside the ` list = {}`. AFAIK there is no command which can be typed after `:` to invoke those actions in nvim-tree
in my original keybinds:
l opens the folder and in case of a file opens it to edit it
h closes the node
and v open the file in a vertical split

**Comment by mehalter** [comment]
@throwaway69420-69420 I just merged a PR today that should fix this. Could you try modifying your override slightly to look like this, the only change is the override table key for nvim tree:

```lua
["nvim-tree"] = {
  hijack_cursor = true,
  hijack_directories = {
    enable = true,
    auto_open = true,
  },
  view = {
    mappings = {
      custom_only = false,
      list = {
        { key = { "l", "<CR>", "o" }, cb = tree_cb "edit" },
        { key = "h", cb = tree_cb "close_node" },
        { key = "v", cb = tree_cb "vsplit" },
      },
    },
  },
},
```

**Comment by throwaway69420-69420** [comment]
Hi @mehalter, I would update you within 24 hours
Thanks

**Comment by abalmos** [comment]
I think the problem is related to calling `require("nvim-tree.config")` in `settings.lua` ... I'm guessing that `nvim-tree` is not loaded yet at this point?

For me, when I have that require in `settings.lua` then nothing from `settings.lua` is respected and AstroVim runs with defaults. I haven't had a chance to chase down an error, but commenting the `nvim-tree.config` require fixes things.

**Comment by mehalter** [comment]
Oh I see now, is the callback necessary? To achieve what you want to do can you use:

```lua
["nvim-tree"] = {
  hijack_cursor = true,
  hijack_directories = {
    enable = true,
    auto_open = true,
  },
  view = {
    mappings = {
      custom_only = false,
      list = {
        { key = { "l", "<CR>", "o" }, action = "edit" },
        { key = "h", action = "close_node" },
        { key = "v", action = "vsplit" },
      },
    },
  },
},
```

I also don't quite understand what the callback part is support to achieve, but the docs on nvim tree show that it may be deprecated (https://github.com/kyazdani42/nvim-tree.lua/blob/08c57660a7761a507a3ae2f4a7c71fe30a412537/doc/nvim-tree-lua.txt#L723)

**Comment by mehalter** [comment]
@abalmos @throwaway69420-69420 I have just opened a new PR that extends this overrides function to be more extensible and cover this use case #136 . Take a look at the example I have included in the PR for `cmp` as it is very similar where it relies on loading the `cmp` plugin.

**Comment by mehalter** [comment]
TL;DR with these changes you would be able to make a file in the `user/` config folder  named `nvim-tree.lua` with the contents:

```
local tree_cb = require("nvim-tree.config").nvim_tree_callback

return {
  hijack_cursor = true,
  hijack_directories = {
    enable = true,
    auto_open = true,
  },
  view = {
    mappings = {
      custom_only = false,
      list = {
        { key = { "l", "<CR>", "o" }, cb = tree_cb "edit" },
        { key = "h", cb = tree_cb "close_node" },
        { key = "v", cb = tree_cb "vsplit" },
      },
    },
  },
}
```

and it will work. These external override files are loaded when the plugin is being configured and so you will be able to access their functions.

**Comment by throwaway69420-69420** [comment]
hi @mehalter , I can't test it now due to some reasons but does this method work for all the default plugins?

**Comment by mehalter** [comment]
@throwaway69420-69420  Yup! this works for all of the default plugins. Another method that you can do to get it to lazy load but stay in the `settings.lua` file is to do a `function(table)` in the `overrides` table. Where the `table` parameter is the default table and you return the table you would actually like to be set. These functions are called when they are needed and so the necessary plugins are available as well. So this could look like this in `settings.lua`:

```lua
  overrides = {
    ["nvim-tree"] = function(table)
      local tree_cb = require("nvim-tree.config").nvim_tree_callback

      local new_opts = {
        hijack_cursor = true,
        hijack_directories = {
          enable = true,
          auto_open = true,
        },
        view = {
          mappings = {
            custom_only = false,
            list = {
              { key = { "l", "<CR>", "o" }, cb = tree_cb "edit" },
              { key = "h", cb = tree_cb "close_node" },
              { key = "v", cb = tree_cb "vsplit" },
            },
          },
        },
      }

      return vim.tbl_deep_extend("force", table, new_opts)
    end,
  },
```

**Comment by mehalter** [comment]
@throwaway69420-69420 let me know when you are able to give this a try and verify it works for you! I'm going to go ahead and close the issue and we can re open it if the problem persists.

**Comment by throwaway69420-69420** [comment]
@mehalter sure, I will let you know as soon as possible
